Refactor certificate documentation - bring your own key ring (#4903)
zowe/docs-site: docs/user-guide/certificates-configuration-scenarios.md

# Certificate configuration scenarios 

After you have completed the [Zowe certificates configuration questionnaire](./certificates-configuration-questionnaire.md) and chosen a certificate configuration scenario, follow the scenario-specific documentation and `zowe.yaml` examples in this article.

:::info Required roles: system programmer, security administrator
:::

Zowe servers require both a keystore to store the certificates and a truststore to validate certificates.

> Zowe's assisted certificate setup begins with customizing certificate scenario YAML files found in the `files/examples/setup/certificate` sub-directory within your Zowe installation directory. Zowe will then use these scenario configurations and complete certificate setup via the `zwe init certificate` command. 

Reviewer: ```suggestion
Zowe's assisted certificate setup begins with customizing certificate scenario YAML files found in the `files/examples/setup/certificate` sub-directory within your Zowe installation directory. Zowe then uses these scenario configurations and complete certificate setup via the `zwe init certificate` command. 
```

:::note 

> Reminder: Zowe assisted generation of certificates is an option, but is not required. If you already have certificates and a key ring or keystore / truststore combination you created on your own, you can instead go to [Finalize certificate configuration](./certificates-finalize-configuration.md).

Reviewer: ```suggestion
Reminder: Zowe assisted generation of certificates is an option, but is not required. If you already have certificates and a key ring or keystore/truststore combination you created on your own, you can instead go to [Finalize certificate configuration](./certificates-finalize-configuration.md).
```

* [Scenario 3: Use a z/OS keyring-based keystore with Zowe generated certificates](#scenario-3-use-a-zos-keyring-based-keystore-with-zowe-generated-certificates)
* [Scenario 4: Use a z/OS keyring-based keystore and connect an existing certificate](#scenario-4-use-a-zos-keyring-based-keystore-and-connect-to-an-existing-certificate)
* [Scenario 5: Use a z/OS keyring-based keystore and import a certificate stored in a data set](#scenario-5-use-a-zos-keyring-based-keystore-and-import-a-certificate-stored-in-a-data-set)

:::note
Ensure that all certificate alias values for all scenarios use only lower-case.
:::

## Running a scenario file

Zowe supplies 5 certificate scenario YAML files in the `files/examples/setup/certificate` sub-directory within your Zowe installation directory. The scenario section below describes what each scenario will accomplish, and how to complete customization of the respective scenario YAML files. Once you have selected and finished configuring one of these scenarios, follow the below instructions to execute the scenario.

<details>
<summary>Run scenario YAML with configmgr</summary>

> 1. Identify the fully-qualified path to your selected scenario YAML configuration file. In this example, we will use `/path/to/files/examples/setup/certificate/scenario-1.yaml`.

Reviewer: ```suggestion
1. Identify the fully-qualified path to your selected scenario YAML configuration file. In this example, we use `/path/to/files/examples/setup/certificate/scenario-1.yaml`.
```

> 2. Identify the fully-qualified path to your main `zowe.yaml` configuration file. In this example, we will use `/path/to/zowe.yaml`.

Reviewer: ```suggestion
2. Identify the fully-qualified path to your main `zowe.yaml` configuration file. In this example, we use `/path/to/zowe.yaml`.
```

    ```
    FILE(/path/to/files/examples/setup/certificate/scenario-1.yaml):FILE(/path/to/zowe.yaml)
    ```

4. Navigate to your Zowe installation root directory. <!-- TODO: do we have an easy way/link to say set zwe on your PATH? -->

4. Submit `./bin/zwe init certificate` using the configuration parameter from Step 3. Note the quotation marks surrounding the configuration parameter! If you are running Scenario 3, 4, or 5, you can add `--dry-run` to the command and review the JCL used to create the keyring before submission.
    ```shell
    ./bin/zwe init certificate --update-config -c "FILE(/path/to/files/examples/setup/certificate/scenario-1.yaml):FILE(/path/to/zowe.yaml)" 

    # If you are using scenarios 3, 4, or 5, this will print JCL to the console for review
    ./bin/zwe init certificate --dry-run --update-config -c "FILE(/path/to/files/examples/setup/certificate/scenario-1.yaml):FILE(/path/to/zowe.yaml)" 
    ```

</details>

<!--<

Does not work right now, and does not work if you try to do something like 

...first yaml
---
# scenario yaml
zowe:
  setup:
    ...

details>
<summary>Append scenario YAML contents to main zowe.yaml</summary>

> 1. Identify the fully-qualified path to your selected scenario YAML configuration file. In this example, we will use `/path/to/files/examples/setup/certificate/scenario-1.yaml`.

Reviewer: ```suggestion
1. Identify the fully-qualified path to your selected scenario YAML configuration file. In this example, we use `/path/to/files/examples/setup/certificate/scenario-1.yaml`.
```

::

</details>

## Scenario 2: Use a file-based (PKCS12) keystore and import a certificate generated by another CA

Follow this procedure and configure `scenario-2.yaml` within the `files/examples/setup/certificate` directory to enable Zowe to use previously generated certificates with a PKCS12 keystore stored in a USS directory.

<details>
<summary>Click here for details.</summary>

1. Set the `type` of the certificate storage to `PKCS12`.
2. Customize the keystore `directory` in the following format:
    ```yaml
    directory: /var/zowe/keystore
    ```
3. Lock the keystore directory so it is accessible only to the Zowe runtime user and group:
    ```yaml
    lock: true
    ```

> 4. Set keystore password. The default value is `password`.

Author: I went with `The value is: xyz` for all uses except sample code blocks. Within the code blocks, I kept it as `The value is xyz` because that matches comments users will see in their zowe.yaml files.

        ```yaml
        importCertificateAuthorities:
          - ""
        ```
:::note
Due to the limitation of `RACDCERT` command, this field should contain a maximum of 2 entries.
::: 

  **Example zowe yaml for scenario 4:**

```yaml
zowe:
  setup:
    certificate:
      type: JCERACFKS
      keyring:
        name: ZoweKeyring
        connect:
          user: IBMUSER
          label: ""
      importCertificateAuthorities:
        - "zOSMFCA"
```

Your yaml file is now configured to enable Zowe to use a z/OS keyring-based keystore with Zowe generated certificates. You can now go to [Running a scenario file](#running-a-scenario-file).

The following example uses an existing JCERACFKS certificate for Zowe's z/OS components. For more information about configuration in this scenario, see [this Medium blog post](https://medium.com/zowe/master-zowe-certificates-use-an-existing-jceracfks-certificate-for-zowes-z-os-components-975ffa0d9f2f), or the video tutorials in [this YouTube playlist](https://www.youtube.com/playlist?list=PL8REpLGaY9QEHLNA81DRgGqWcgOYC0PDX).

**Example:**
![Alt text](../images/certificates/connect-JCERACFKS.png)

**Note:**

> In this example, the command `zwe init certificate -c ./zowe.yaml --security-dry-run` allows the JCL to be inspected before submission, as well as handed off to a security administrator who has privileges to submit the JCL under their user ID. By default, the JCL id submitted immediately. 

Reviewer: ```suggestion
In this example, the command `zwe init certificate -c ./zowe.yaml --security-dry-run` allows the JCL to be inspected before submission, as well as handed off to a security administrator who has privileges to submit the JCL under their user ID. By default, the JCL is submitted immediately. 
```

    ```yaml
    lock: true
    ```
5. (Optional) Set the alias name of certificate authority. The default value is: `local_ca`.  
   ```yaml
   caAlias: local_ca
   ```
6. (Optional) Set the password of the keystore stored certificate authority. The default value is: `local_ca_password`.
    ```yaml
    caPassword: local_ca_password
    ```
7. Set the validity in days for the Zowe generated certificates
    ```yaml  
    validity: 3650
    ```
8. (Optional) Specify the distinguished name for Zowe generated certificates. If you do not want to customize this, you can leave it commented out.
    ```yaml
    dname:
      caCommonName: ""
      commonName: ""
      orgUnit: ""
      org: ""
      locality: ""
      state: ""
      country: ""
    ```
9. (Optional) Set the domain names and IPs specified in nested subsection `SAN`. If this field is not defined, the `zwe init` command uses the value `zowe.externalDomains`, which are found in your main `zowe.yaml` file.
    ```yaml
    san:
        # sample domain name
        - dvipa.my-company.com
        # sample IP address
        - 12.34.56.78
     ```
:::tip

> To get the san IP address, run `ping dvipa.my-company.com` in your terminal. 

Author: Normalized on `san`.
